File: Python3/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s):
            b = 0
            maxlength = 0
            d = {}
            for i, item in enumerate(s):
                if item in d:
                    b = max(b, d[item] + 1)
                d[item] = i
                maxlength = max(maxlength, i-b+1)
            return maxlength
        return 0

 # @lc code=end

File: C#/3.longest-substring-without-repeating-characters.cs
/*
 * @lc app=leetcode id=3 lang=csharp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
using System.Collections;
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        HashSet<object> notdup = new HashSet<object>();
        Queue anslist = new Queue();
        int maxlen = 0;
        for(int i = 0;i < s.Length;i++){
            if(notdup.Contains(s[i])){
                while(anslist.Count > 0 && notdup.Contains(s[i])){
                    var curele = anslist.Dequeue();
                    notdup.Remove(curele);
                }
            }
            notdup.Add(s[i]);
            anslist.Enqueue(s[i]);
            int curlen = anslist.Count;
            maxlen = curlen > maxlen ? curlen : maxlen;
        }
        return maxlen;
    }
}
// @lc code=end

